core/state.go

```diff
@@ -362,39 +352,97 @@ func (s *State) Class(classHash *felt.Felt) (*DeclaredClass, error) {
-// contract at the given address in the given Txn context.
-func (s *State) updateContractStorage(stateTrie *trie.Trie, addr *felt.Felt, diff []StorageDiff, onChanged OnValueChanged,
-	blockNumber uint64,
-) error {
-	contract, err := NewContract(addr, s.txn)
+func (s *State) updateStorageBuffered(contractAddr *felt.Felt, updateDiff []StorageDiff, blockNumber uint64, logChanges bool) (
+	*db.BufferedTransaction, error,
+) {
+	// to avoid multiple transactions writing to s.txn, create a buffered transaction and use that in the worker goroutine
+	bufferedTxn := db.NewBufferedTransaction(s.txn)
+	bufferedState := NewState(bufferedTxn)
+	bufferedContract, err := NewContract(contractAddr, bufferedTxn)
 	if err != nil {
-		if !errors.Is(err, ErrContractNotDeployed) {
-			return err
+		return nil, err
+	}
+
+	onValueChanged := func(location, oldValue *felt.Felt) error {
+		if logChanges {
+			return bufferedState.LogContractStorage(contractAddr, location, oldValue, blockNumber)
 		}
+		return nil
+	}
 
-		if _, ok := noClassContracts[*addr]; !ok {
-			return err
+	if err = bufferedContract.UpdateStorage(updateDiff, onValueChanged); err != nil {
+		return nil, err
+	}
+
+	return bufferedTxn, nil
+}
+
+// updateContractStorage applies the diff set to the Trie of the
+// contract at the given address in the given Txn context.
+func (s *State) updateContractStorages(stateTrie *trie.Trie, diffs map[felt.Felt][]StorageDiff, blockNumber uint64, logChanges bool) error {
+	// make sure all noClassContracts are deployed
+	for addr := range diffs {
+		if _, ok := noClassContracts[addr]; !ok {
+			continue
 		}
 
-		// Deploy noClassContract
-		err = s.putNewContract(stateTrie, addr, noClassContractsClassHash, blockNumber)
+		_, err := NewContract(&addr, s.txn)
 		if err != nil {
+			if !errors.Is(err, ErrContractNotDeployed) {
+				return err
+			}
+			// Deploy noClassContract
+			err = s.putNewContract(stateTrie, &addr, noClassContractsClassHash, blockNumber)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// sort the contracts in decending diff size order
+	// so we start with the heaviest update first
+	keys := make([]felt.Felt, 0, len(diffs))
+	for key := range diffs {
+		keys = append(keys, key)
+	}
+	sort.SliceStable(keys, func(i, j int) bool {
+		return len(diffs[keys[i]]) > len(diffs[keys[j]])
+	})
+
+	// update per-contract storage Tries concurrently
+	contractUpdaters := pool.NewWithResults[*db.BufferedTransaction]().WithErrors().WithMaxGoroutines(runtime.GOMAXPROCS(0))
+	for _, key := range keys {
+		conractAddr := key
+		updateDiff := diffs[conractAddr]
+		contractUpdaters.Go(func() (*db.BufferedTransaction, error) {
+			return s.updateStorageBuffered(&conractAddr, updateDiff, blockNumber, logChanges)
+		})
+	}
+
+	bufferedTxns, err := contractUpdaters.Wait()
+	if err != nil {
+		return err
+	}
+
+	// flush buffered txns
+	for _, bufferedTxn := range bufferedTxns {
+		if err = bufferedTxn.Flush(); err != nil {
 			return err
 		}
+	}
```

Comment:
Actually if we sort the slice of BufferedTransactions we get back from the routines in ascending contract address order, we might get an additional speed up. But it is not really worth it at this point. Will leave it to another PR.